Source/EasyNetQ/Consumer/ConsumerDispatcher.cs

```diff
@@ -8,53 +8,73 @@ namespace EasyNetQ.Consumer
     public class ConsumerDispatcher : IConsumerDispatcher
     {
         private readonly ILog logger = LogProvider.For<ConsumerDispatcher>();
-        private readonly BlockingCollection<Action> queue;
-        private bool disposed;
+        private readonly AutoResetEvent autoResetEvent = new AutoResetEvent(false);
+        private readonly ConcurrentQueue<Action> durableActions = new ConcurrentQueue<Action>();
+        private readonly ConcurrentQueue<Action> transientActions = new ConcurrentQueue<Action>();
 
         public ConsumerDispatcher(ConnectionConfiguration configuration)
         {
             Preconditions.CheckNotNull(configuration, "configuration");
 
-            queue = new BlockingCollection<Action>();
-
             var thread = new Thread(_ =>
             {
-                while (!disposed && queue.TryTake(out var action, -1))
-                {
+                while (!IsDone())
                     try
                     {
-                        action();
+                        if (durableActions.TryDequeue(out var action) || transientActions.TryDequeue(out action))
+                            action();
+                        else
+                            autoResetEvent.WaitOne();
                     }
                     catch (Exception exception)
                     {
                         logger.ErrorException(string.Empty, exception);
                     }
-                }
             }) {Name = "EasyNetQ consumer dispatch thread", IsBackground = configuration.UseBackgroundThreads};
             thread.Start();
         }
 
+        public bool IsDisposed { get; private set; }
```

Comment:
Why this property is public?

Also I seems we need a backing field and to mark it volatile to make changes visible, because it will be accessed(dispatch thread) and modified(other thread which wants to dispose) from different threads.

Reply:
it is used in a test in ConsumerDispatcherFactoryTests "Should_dispose_dispatcher_when_factory_is_disposed". I think the test should be removed and IsDisposed to a Volatile backing field :)

Reply:
I missed the usage, sorry

Reply:
Agree, let's remove the test.